pkg/rc/replication_controller.go

```diff
@@ -141,72 +157,129 @@ func (rc *replicationController) meetDesires() error {
+				rc.ReplicasDesired, len(currentNodes), len(eligible), i,
+			)
 		}
-	} else if len(current) > rc.ReplicasDesired {
-		eligible, err := rc.eligibleNodes()
+
+		err := rc.schedule(scheduleOn)
 		if err != nil {
 			return err
 		}
+	}
+	return nil
+}
 
-		// If we need to downsize the number of nodes, prefer any in current that are not eligible anymore.
-		// TODO: evaluate changes to 'eligible' more frequently
-		preferred := sets.NewString(current...).Difference(sets.NewString(eligible...))
-		rest := sets.NewString(current...).Difference(preferred)
-		toUnschedule := len(current) - rc.ReplicasDesired
-		rc.logger.NoFields().Infof("Need to unschedule %d nodes out of %s", toUnschedule, current)
+func (rc *replicationController) removePods(current PodLocations) error {
+	currentNodes := current.Nodes()
+	eligible, err := rc.eligibleNodes()
+	if err != nil {
+		return err
+	}
 
-		for i := 0; i < toUnschedule; i++ {
-			unscheduleFrom, ok := preferred.PopAny()
+	// If we need to downsize the number of nodes, prefer any in current that are not eligible anymore.
+	// TODO: evaluate changes to 'eligible' more frequently
+	preferred := sets.NewString(currentNodes...).Difference(sets.NewString(eligible...))
+	rest := sets.NewString(currentNodes...).Difference(preferred)
+	toUnschedule := len(current) - rc.ReplicasDesired
+	rc.logger.NoFields().Infof("Need to unschedule %d nodes out of %s", toUnschedule, current)
+
+	for i := 0; i < toUnschedule; i++ {
+		unscheduleFrom, ok := preferred.PopAny()
+		if !ok {
+			var ok bool
+			unscheduleFrom, ok = rest.PopAny()
 			if !ok {
-				var ok bool
-				unscheduleFrom, ok = rest.PopAny()
-				if !ok {
-					// This should be mathematically impossible unless replicasDesired was negative
-					return util.Errorf(
-						"Unable to unschedule enough nodes to meet replicas desired: %d replicas desired, %d current.",
-						rc.ReplicasDesired, len(current),
-					)
-				}
+				// This should be mathematically impossible unless replicasDesired was negative
+				return util.Errorf(
+					"Unable to unschedule enough nodes to meet replicas desired: %d replicas desired, %d current.",
+					rc.ReplicasDesired, len(current),
+				)
 			}
-			err := rc.unschedule(unscheduleFrom)
+		}
+		err := rc.unschedule(unscheduleFrom)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (rc *replicationController) ensureConsistency(current PodLocations) error {
+	manifestSHA, err := rc.Manifest.SHA()
+	if err != nil {
+		return err
+	}
+	for _, pod := range current {
+		intent, _, err := rc.kpStore.Pod(kp.INTENT_TREE, pod.Node, types.PodID(pod.PodID))
+		if err != nil && err != pods.NoCurrentManifest {
+			return err
+		}
+		var intentSHA string
+		if intent != nil {
+			intentSHA, err = intent.SHA()
 			if err != nil {
-				return err
+				rc.logger.WithError(err).WithField("node", pod.Node).Warn("Could not hash manifest to determine consistency of intent")
```

Comment:
I am utterly unable to replicate that problem, even when I check out commit 7181aa8, the parent of 8edd6cb where `WithError()` was introduced. The following program fragment:

```
func main() {
	logging.DefaultLogger.WithError(util.Errorf("This is an error")).
        WithField("foo", "foo").WithField("bar", "bar").Infoln("My message")
}
```

always produces output with keys `"err"`, `"filename"`, `"function"`, `"line_number"`, `"foo"`, `"bar"`, et al., just as I would anticipate. Perhaps @mpuncel could provide more context about what was broken back then.